### ISSUE TITLE
Detecting codepoint readiness

### ISSUE THREAD

**Issue by garretrieger** [w3c/IFT#223]
This is a sub issue of #201. In general the client will need to be able to detect at any point which codepoints/features/design space combintations in an IFT font are ready to use so that the client can apply font fallback strategies if the unit is not ready yet. I've thought about some different approaches to how this could be done and came up with the following approach:

- Given a span of text that the client is looking to use an IFT font for it can assign each code point in that span the specific set of features being used and the specific point in design space that the code point will be rendered with. Note: here we're looking at the ordered set of code points, so the same codepoint may appear multiple times with different feature and design space configurations.
-  In patch mappings individual code points match patch entries completely independent of other code points, since we use set intersection when checking entries.
- Using the above two points the client can check for each code point in the span if there are any unapplied entries which would match that code point + feature + design space combination.
- If one more more unapplied entries matches then the client will consider that code point to not be ready and apply whatever usual fallback strategies it would normally use. Otherwise the code point can be rendered with the IFT font.
- I believe this should be easy to incorporate into browser's existing fallback handling. For example from what I know of chromes handling of unicode range fallback this would fit in pretty naturally.

We can incorporate this into the spec by adding an additional section which provides an algorithm clients can use to determine readiness.

This appears to me to handle the general case of checking readiness both for the initial font and at any point during the augmentation process for a font. Unfortunately however, it will no longer work if we adopt the new mechanism proposed in #222. This is because that PR would make patch selection no longer independent per codepoint. For example consider the case where data is already loaded for codepoint a and it's considered ready but then later on a UVS selector gets added to the text used with codepoint a where the variant glyph for a isn't yet loaded. To be correct we wouldn't want to consider (UVS + a) ready, but that requires the client considering combinations of codepoints in the input text span when checking readiness. Which in turn requires understanding how codepoints interact inside of the font. Codepoints don't always need to be adjacent to interact. Thus it isn't always possible without having a fully loaded the relevant font data, since the patch mapping doesn't capture how the codepoints interact only that they do in some way.

As a result, I think we may want to consider not adopting the mechanism proposed in #222 because it will make determining readiness in the browser far more difficult and complicated. We are unfortunately forced to make a tradeoff here and reduce our ability to optimally encode certain cases like UVS selectors, but ultimately I think it's more important that the client have a clean and easy way to detect readiness on a per codepoint basis (to align with existing fallback handling).

I plan to spend a bit more time reviewing chromes fallback handling just to make sure the approach proposed above actually would integrate well.

**Comment by skef** [comment]
> In general the client will need to be able to detect at any point which codepoints/features/design space combintations in an IFT font are ready to use so that the client can apply font fallback strategies if the unit is not ready yet.

I've been mentally working from a subset-parameter "readiness" model, where you have a set of codepoints, layout feature tags, and a range of axis locations, and you ask "are the patches relative to this set fully loaded?" (or maybe "sufficiently loaded"). 

It sounds like you think it's not going to work to ask the question at that "grain". Or maybe that you'd rather not force the client to think in those terms.

**Comment by garretrieger** [comment]
> > In general the client will need to be able to detect at any point which codepoints/features/design space combintations in an IFT font are ready to use so that the client can apply font fallback strategies if the unit is not ready yet.
> 
> I've been mentally working from a subset-parameter "readiness" model, where you have a set of codepoints, layout feature tags, and a range of axis locations, and you ask "are the patches relative to this set fully loaded?" (or maybe "sufficiently loaded").
> 
> It sounds like you think it's not going to work to ask the question at that "grain". Or maybe that you'd rather not force the client to think in those terms.

A good example of where that approach may run into issues is an IFT font which has optional small caps feature. Let's assume the font's non small caps glyphs are divided into a set of patches, and there's a second set of patches with the small cap glyphs which requires the smcp feature to be present.

To start with you have some text not using small caps so the client loads the relevant non small caps bins. Now the client can look at the non small caps text see that those characters don't have any outstanding patches (since the small caps feature isn't activated on this text, the small caps patches won't be matched) and conclude it is safe to use the font for them.

Then let's say an additional word gets added to the content, this time with the small caps featured applied (only to the new word). This now causes all of the small caps patches which intersect all of the content seen so far to be activated. If you look at readiness only from the subset parameter point of view, it now looks like the the font can't be used for the whole string of text. Which in the browser case would cause it to flip that text back to a fallback font. In reality though, just the newly added word using the small caps feature is the only part that the font can't be used for yet.

Following this leads to the model I proposed above where the client checks each codepoint + design space/features applied to it in isolation against the set of unapplied patches. Another use case where checking codepoint by codepoint is useful is handling the case where patches are partially loaded. For example the client has requested patches for several of the small cap bins, but at the moment some but not all of them have been received and applied.

**Comment by skef** [comment]
> Following this leads to the model I proposed above where the client checks each codepoint + design space/features applied to it in isolation against the set of unapplied patches. Another use case where checking codepoint by codepoint is useful is handling the case where patches are partially loaded. For example the client has requested patches for several of the small cap bins, but at the moment some but not all of them have been received and applied.

Taking "codepoint by codepoint" literally, it seems like this wouldn't even work for, e.g., `dlig`.

Have we discussed this before? It seems like a potentially large and complex area that's suddenly opened up.

My initial instinct would be to say that if a client encounters a long string that it needs to fall back on, it would be OK to try to break it into word boundaries (assuming "words" can be well defined) and check those against full parameter sets but it can't go lower than that. In places were words can't be well defined -- suck it up and temporarily fall back, even at the cost of some flashing.

**Comment by skef** [comment]
I guess given the current specification `dlig` would work correctly.

However, I'm really iffy on this in general.

We've been very vague and conceptual on the requirements, for example the closure requirement for glyph-keyed fonts. Basically our guidance has been: "just make sure things work out right". This was more or less insisted on. 

Now we have this quite different codepoint-by-codepoint assessment. And I'm very doubtful this can be limited to *temporary* questions of "readiness" -- presumably if you determine *everything* you need to render can be rendered correctly with current patches, then you don't need to load any more patches. Right? So now there's a quite different standard by which things need to "work out". Which am I encoding to, the closure requirement for the codepoint-by-codepoint requirement?

**Comment by garretrieger** [comment]
Follow up on our discussion from today's WG call:
- Skef brought up a case where there is a contextual substitution (eg. abc -> aBc) which is enabled by an optional feature. If the feature is only enabled for 'b' shapers may still allow the rule to activate. If true this means the per code point readiness check I outlined above is not always accurate. 
- I checked how harfbuzz handles this case and there are cases (eg. chain context substitution rules) which can activate a substitution even if the context characters do not have the feature activated.

Which means for this proposal codepoint by codepoint readiness checking is not accurate in all cases, in the example above it could declare "b" ready to use with the IFT font even if the substitution rule has not yet been loaded.

This also leads into a related issue of the granularity of the subset definition that the client is using for extension.  The specification is currently silent on how say a page of content is transformed into a subset definition for the loading process. For example is all of the content in the page using the font unioned together into one large definition, or is the client allowed to keep separate subset definitions for distinctly styled spans of text?

Based on the above I believe there's a few changes we'll want to make:
1. Add some clarification text regarding the input subset definition and require that at minimum it must be on a complete shaping unit as defined by the clients text processing algorithm (normative). For example in Chrome this would typically be a word since chrome runs the shaper on one word at a time. This combined with our existing encoder requirements will ensure correct loading and subsequent rendering.
2. Add a non-normative section that provides a recommended readiness check that clients can opt to use (but also free to implement their own). We probably want a two tiered one: first readiness is checked for each minimum shaping unit (this is always correct by point 1), and then if a shaping unit fails the readiness check a second approximate approach is used within the unit. For example checking graphemes, individual codepoints, or Skef's suggestion of scanning forward in the words writing direction until the check fails. Make it clear the second tier (within a shaping unit) is an approximation and may incorrectly report readiness in certain complex cases.
3. Make it clear that the readiness check in 2 should not influence patch selection and loading decisions. Per .1 those must always be based at minimum on a complete shaping unit.

As a result of the above an encoder only needs to reason about the encoding in terms of point 1, there will be no requirement that they make encoding decisions with respect to readiness checking clients might do.

This also gives us a way to move forward with the UVS fix in #222 since the first tier of readiness checking will work correctly even under the proposed changes and any failures in the approximate tier will only have effect within a single shaping unit. Also the approximate strategies should be correct for the majority of typical cases, so we would expect pretty small number of occurrences of incorrect readiness decisions in practice.
